### PR TITLE
GP - 160: create a completed solution for Enable-string-trimming exer…

### PR DESCRIPTION
**Pull request by Stanislav-Zabramnyi**
…cise

### CODE REVIEW

File: 3-0-spring-framework/3-3-0-enable-string-trimming/src/main/java/com/bobocode/StringTrimmingConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,11 @@
+package com.bobocode;
+
+import org.springframework.context.annotation.Bean;
+
+public class StringTrimmingConfiguration {
```
Comment (tboychuk) [inline 3-0-spring-framework/3-3-0-enable-string-trimming/src/main/java/com/bobocode/StringTrimmingConfiguration.java:5]: @Stanislav-Zabramnyi please add javadoc (explain that it's porously not marked with `@Configuration`)

File: 3-0-spring-framework/3-3-0-enable-string-trimming/src/main/java/com/bobocode/TrimmedAnnotationBeanPostProcessor.java [hunk 2]
```diff
@@ -1,21 +1,106 @@
...
+import org.springframework.util.StringUtils;
 
 /**
  * This is processor class implements {@link BeanPostProcessor}, looks for a beans where method parameters are marked with
  * {@link Trimmed} annotation, creates proxy of them, overrides methods and trims all {@link String} arguments marked with
  * {@link Trimmed}. For example if there is a string " Java   " as an input parameter it has to be automatically trimmed to "Java"
  * if parameter is marked with {@link Trimmed} annotation.
  * <p>
- *
+ * <p>
```
Comment (tboychuk) [inline 3-0-spring-framework/3-3-0-enable-string-trimming/src/main/java/com/bobocode/TrimmedAnnotationBeanPostProcessor.java:26]: @Stanislav-Zabramnyi such changes should be a part of `main`

File: 3-0-spring-framework/3-3-0-enable-string-trimming/src/main/java/com/bobocode/TrimmedAnnotationBeanPostProcessor.java [hunk 3]
```diff
@@ -1,21 +1,106 @@
...
+    private final Map<String, Object> beansToBeProxied = new HashMap<>();
+    private final Map<String, MethodInterceptor[]> beanInterceptors = new HashMap<>();
+
+    @Override
+    public Object postProcessBeforeInitialization(Object bean, String beanName) throws BeansException {
+        if (containsParametersAnnotatedWithTrimmed(bean)) {
+            beansToBeProxied.put(beanName, bean);
+            beanInterceptors.put(beanName, new MethodInterceptor[]{provideTrimmingInterceptor()});
+        }
+        return BeanPostProcessor.super.postProcessBeforeInitialization(bean, beanName);
```
Comment (tboychuk) [inline 3-0-spring-framework/3-3-0-enable-string-trimming/src/main/java/com/bobocode/TrimmedAnnotationBeanPostProcessor.java:42]: @Stanislav-Zabramnyi you can go with `super.postProcessBeforeInitialization(bean, beanName);`

File: 3-0-spring-framework/3-3-0-enable-string-trimming/src/main/java/com/bobocode/TrimmedAnnotationBeanPostProcessor.java [hunk 4]
```diff
@@ -1,21 +1,106 @@
...
+ * <p>
  * Note! This bean is not marked as a {@link Component} to avoid automatic scanning, instead it should be created in
  * {@link StringTrimmingConfiguration} class which can be imported to a {@link Configuration} class by annotation
  * {@link EnableStringTrimming}
  */
-public class TrimmedAnnotationBeanPostProcessor {
-//todo: Implement TrimmedAnnotationBeanPostProcessor according to javadoc
+public class TrimmedAnnotationBeanPostProcessor implements BeanPostProcessor {
+
+    private final Map<String, Object> beansToBeProxied = new HashMap<>();
```
Comment (tboychuk) [inline 3-0-spring-framework/3-3-0-enable-string-trimming/src/main/java/com/bobocode/TrimmedAnnotationBeanPostProcessor.java:33]: @Stanislav-Zabramnyi you don't need to store a bean, only it's class. So it should be
`private final Map<String, Class<?>> classesToBeProxied = new HashMap<>();`

File: 3-0-spring-framework/3-3-0-enable-string-trimming/src/main/java/com/bobocode/TrimmedAnnotationBeanPostProcessor.java [hunk 5]
```diff
@@ -1,21 +1,106 @@
...
+
+    private MethodInterceptor provideTrimmingInterceptor() {
+        return (beanInstance, method, args, methodProxy) ->
+                methodProxy.invokeSuper(beanInstance, processParameters(method, args));
+    }
+
+    private Object[] processParameters(Method method, Object[] args) {
+        Parameter[] parameters = method.getParameters();
+        for (int i = 0; i < parameters.length; i++) {
+            if (parameters[i].isAnnotationPresent(Trimmed.class)) {
```
Comment (tboychuk) [inline 3-0-spring-framework/3-3-0-enable-string-trimming/src/main/java/com/bobocode/TrimmedAnnotationBeanPostProcessor.java:65]: @Stanislav-Zabramnyi so are you sure that this `Method` is not a method of a proxy class?

File: 3-0-spring-framework/3-3-0-enable-string-trimming/src/main/java/com/bobocode/TrimmedAnnotationBeanPostProcessor.java [hunk 6]
```diff
@@ -1,21 +1,106 @@
...
+            enhancer.setCallbacks(methodInterceptors);
+        }
+    }
+
+    @SneakyThrows
+    private MethodInterceptor[] findInterceptors(Object bean) {
+        var method = Arrays.stream(bean.getClass().getDeclaredMethods()).filter(m -> m.getName().equals("getCallbacks"))
+                .findFirst();
+        if (method.isPresent()) {
+            return (MethodInterceptor[]) method.get().invoke(bean);
```
Comment (tboychuk) [inline 3-0-spring-framework/3-3-0-enable-string-trimming/src/main/java/com/bobocode/TrimmedAnnotationBeanPostProcessor.java:102]: @Stanislav-Zabramnyi it looks scary! 😳

File: 3-0-spring-framework/3-3-0-enable-string-trimming/src/main/java/com/bobocode/TrimmedAnnotationBeanPostProcessor.java [hunk 7]
```diff
@@ -1,21 +1,106 @@
...
+        return arg;
+    }
+
+    private boolean containsParametersAnnotatedWithTrimmed(Object bean) {
+        return Arrays.stream(bean.getClass().getDeclaredMethods()).flatMap(m -> Stream.of(m.getParameters()))
+                .anyMatch(p -> p.isAnnotationPresent(Trimmed.class));
+    }
+
+    private void setInterceptorsIfNeeded(Enhancer enhancer, Object bean, String beanName) {
+        MethodInterceptor[] methodInterceptors = beanInterceptors.get(beanName);
```
Comment (tboychuk) [inline 3-0-spring-framework/3-3-0-enable-string-trimming/src/main/java/com/bobocode/TrimmedAnnotationBeanPostProcessor.java:85]: @Stanislav-Zabramnyi to me **this second map `beanInterceptors` is redundant**. The only thing it's storing is an interceptor  created by `provideTrimmingInterceptor`. So we should get rid of that map, and just call `provideTrimmingInterceptor` when we need it.

File: 3-0-spring-framework/3-3-0-enable-string-trimming/src/main/java/com/bobocode/TrimmedAnnotationBeanPostProcessor.java [hunk 8]
```diff
@@ -1,21 +1,106 @@
...
+        Parameter[] parameters = method.getParameters();
+        for (int i = 0; i < parameters.length; i++) {
+            if (parameters[i].isAnnotationPresent(Trimmed.class)) {
+                args[i] = trimMarkedString(args[i]);
+            }
+        }
+        return args;
+    }
+
+    private Object trimMarkedString(Object arg) {
```
Comment (tboychuk) [inline 3-0-spring-framework/3-3-0-enable-string-trimming/src/main/java/com/bobocode/TrimmedAnnotationBeanPostProcessor.java:72]: @Stanislav-Zabramnyi I would rather say 
```java
        if (arg != null && arg instanceof String string) {
            return string.trim();
        }
        return arg;
```

**Review by tboychuk — COMMENTED** [review]
@Stanislav-Zabramnyi thanks for the impl. Now we need to talk. I'd like to know more about those interceptors :)

File: 3-0-spring-framework/3-3-0-enable-string-trimming/src/main/java/com/bobocode/TrimmedAnnotationBeanPostProcessor.java (hunk 3 above)
Comment (Stanislav-Zabramnyi) [reply 3-0-spring-framework/3-3-0-enable-string-trimming/src/main/java/com/bobocode/TrimmedAnnotationBeanPostProcessor.java:42]: sure, this is just a default IDEA overriding

File: 3-0-spring-framework/3-3-0-enable-string-trimming/src/main/java/com/bobocode/TrimmedAnnotationBeanPostProcessor.java (hunk 5 above)
Comment (Stanislav-Zabramnyi) [reply 3-0-spring-framework/3-3-0-enable-string-trimming/src/main/java/com/bobocode/TrimmedAnnotationBeanPostProcessor.java:65]: A method of a proxy class has type `MethodProxy` but not a `Method`.

File: 3-0-spring-framework/3-3-0-enable-string-trimming/src/main/java/com/bobocode/TrimmedAnnotationBeanPostProcessor.java (hunk 8 above)
Comment (Stanislav-Zabramnyi) [reply 3-0-spring-framework/3-3-0-enable-string-trimming/src/main/java/com/bobocode/TrimmedAnnotationBeanPostProcessor.java:72]: condition `arg! = null` has already covered by `arg instanceof String`, if `arg is null` , `arg instanceof String` will just return `false`

File: 3-0-spring-framework/3-3-0-enable-string-trimming/src/main/java/com/bobocode/TrimmedAnnotationBeanPostProcessor.java (hunk 6 above)
Comment (Stanislav-Zabramnyi) [reply 3-0-spring-framework/3-3-0-enable-string-trimming/src/main/java/com/bobocode/TrimmedAnnotationBeanPostProcessor.java:102]: agree. But It`s reflection))  

I will think how to do it in more elegant way.

**Review by tboychuk — APPROVED** [review]
Thanks 👍

**Pull request merged**